{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os \n",
    "import zipfile\n",
    "import StringIO\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import itertools\n",
    "pd.set_option('max_colwidth', 800)\n",
    "\n",
    "datadir = '/Users/kanaaax/Google Drive/TS-EUROTRAIN/RESULTS_QSMv3/dataframes/AHBA/ahba_data/'\n",
    "\n",
    "# https://human.brain-map.org/\n",
    "subjects = ['H0351.2001', 'H0351.2002', 'H0351.1009', 'H0351.1012', 'H0351.1015', 'H0351.1016']\n",
    "donors   = ['donor1'    , 'donor2'    , 'donor3'    , 'donor4'    , 'donor5'    , 'donor6'    ]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Allen Human Brain Atlas microarray data processing\n",
    "\n",
    "- Data:\n",
    "    - Complete normalized microarray datasets of six brains were downloaded from  http://human.brain-map.org/static/download\n",
    "    - The datasets contain gene expression values normalized across all brains using an improved normalization process implemented in March 2013.\n",
    "    - \n",
    "    - Hawrylycz MJ, et al. An anatomically comprehensive atlas of the adult human brain transcriptome. Nature 2012; 489: 391–399.\n",
    "\n",
    "- Processing\n",
    "    - \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "##########################################################################\n",
    "\n",
    "# Function to create 3702x20737 (samples x genes) matrix\n",
    "\n",
    "##########################################################################\n",
    "\n",
    "def return_donor_geneexpression(donor):\n",
    "    \n",
    "    #I/O\n",
    "    print 'Donor:', donor\n",
    "    donor_zip  = zipfile.ZipFile(os.path.join(datadir,'%s.zip'%donor))\n",
    "    \n",
    "    #read microarray data. take care indexing.... should have 58692 probes\n",
    "    df_microarray = pd.read_csv(StringIO.StringIO(donor_zip.read('MicroarrayExpression.csv')),\n",
    "                                header= None,index_col =0 )\n",
    "    print '- Microarray dataframe shape: ',df_microarray.shape\n",
    "\n",
    "    # read probe meta-data\n",
    "    df_probes = pd.read_csv(StringIO.StringIO(donor_zip.read('Probes.csv')), index_col =0 )\n",
    "    del df_probes.index.name\n",
    "    \n",
    "    # drop unannotated genes... should end up with 48170 probes\n",
    "    drop_strings = ['uncharacterized', 'non-RefSeq']\n",
    "    drop_strings = ['non-RefSeq']\n",
    "    drop_probes  = [i for i in df_probes.index for x in drop_strings if x in df_probes.loc[i]['gene_name'] ]\n",
    "    df_probes = df_probes.drop(drop_probes, axis =0)\n",
    "    print '- Probes dataframe shape: ',df_probes.shape\n",
    "    \n",
    "    # Calc mean probe expression for each gene\n",
    "    df_gene_exp = pd.concat([df_microarray, df_probes['gene_symbol']],axis=1).groupby('gene_symbol').mean().T.reset_index().drop(['index'],axis=1)\n",
    "    df_gene_exp.columns.name = None\n",
    "    print '- Gene-expression original-annotation dataframe shape: ',df_gene_exp.shape\n",
    "\n",
    "    # Add metadata to dataframe\n",
    "    df_meta = pd.read_csv(StringIO.StringIO(donor_zip.read('SampleAnnot.csv')))\n",
    "    print '- Metadata dataframe shape: ',df_meta.shape\n",
    "    \n",
    "    df = pd.concat([df_gene_exp,df_meta],axis=1).set_index('well_id')\n",
    "    print '- Full dataframe shape:', df.shape\n",
    "    print ''\n",
    "    del df.index.name\n",
    "    \n",
    "    # append donor name\n",
    "    df['donor'] = donor\n",
    "    return df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Donor: donor1\n",
      "- Microarray dataframe shape:  (58692, 946)\n",
      "- Probes dataframe shape:  (48170, 6)\n",
      "- Gene-expression original-annotation dataframe shape:  (946, 20737)\n",
      "- Metadata dataframe shape:  (946, 13)\n",
      "- Full dataframe shape: (946, 20749)\n",
      "\n",
      "Donor: donor2\n",
      "- Microarray dataframe shape:  (58692, 893)\n",
      "- Probes dataframe shape:  (48170, 6)\n",
      "- Gene-expression original-annotation dataframe shape:  (893, 20737)\n",
      "- Metadata dataframe shape:  (893, 13)\n",
      "- Full dataframe shape: (893, 20749)\n",
      "\n",
      "Donor: donor3\n",
      "- Microarray dataframe shape:  (58692, 363)\n",
      "- Probes dataframe shape:  (48170, 6)\n",
      "- Gene-expression original-annotation dataframe shape:  (363, 20737)\n",
      "- Metadata dataframe shape:  (363, 13)\n",
      "- Full dataframe shape: (363, 20749)\n",
      "\n",
      "Donor: donor4\n",
      "- Microarray dataframe shape:  (58692, 529)\n",
      "- Probes dataframe shape:  (48170, 6)\n",
      "- Gene-expression original-annotation dataframe shape:  (529, 20737)\n",
      "- Metadata dataframe shape:  (529, 13)\n",
      "- Full dataframe shape: (529, 20749)\n",
      "\n",
      "Donor: donor5\n",
      "- Microarray dataframe shape:  (58692, 470)\n",
      "- Probes dataframe shape:  (48170, 6)\n",
      "- Gene-expression original-annotation dataframe shape:  (470, 20737)\n",
      "- Metadata dataframe shape:  (470, 13)\n",
      "- Full dataframe shape: (470, 20749)\n",
      "\n",
      "Donor: donor6\n"
     ]
    },
    {
     "ename": "IOError",
     "evalue": "[Errno 2] No such file or directory: '/Users/kanaaax/Google Drive/TS-EUROTRAIN/RESULTS_QSMv3/dataframes/AHBA/ahba_data/donor6.zip'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mIOError\u001b[0m                                   Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-6-eae80d62e130>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mahba_all\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mdonor\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mdonors\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m     \u001b[0mdf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mreturn_donor_geneexpression\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdonor\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m     \u001b[0mahba_all\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-5-ad5af48ac26f>\u001b[0m in \u001b[0;36mreturn_donor_geneexpression\u001b[0;34m(donor)\u001b[0m\n\u001b[1;32m      9\u001b[0m     \u001b[0;31m#I/O\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     10\u001b[0m     \u001b[0;32mprint\u001b[0m \u001b[0;34m'Donor:'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdonor\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 11\u001b[0;31m     \u001b[0mdonor_zip\u001b[0m  \u001b[0;34m=\u001b[0m \u001b[0mzipfile\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mZipFile\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdatadir\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m'%s.zip'\u001b[0m\u001b[0;34m%\u001b[0m\u001b[0mdonor\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     12\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     13\u001b[0m     \u001b[0;31m#read microarray data. take care indexing.... should have 58692 probes\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/kanaaax/anaconda2/lib/python2.7/zipfile.pyc\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, file, mode, compression, allowZip64)\u001b[0m\n\u001b[1;32m    777\u001b[0m             \u001b[0mmodeDict\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m{\u001b[0m\u001b[0;34m'r'\u001b[0m \u001b[0;34m:\u001b[0m \u001b[0;34m'rb'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'w'\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;34m'wb'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'a'\u001b[0m \u001b[0;34m:\u001b[0m \u001b[0;34m'r+b'\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    778\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 779\u001b[0;31m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfp\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfile\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmodeDict\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mmode\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    780\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mIOError\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    781\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0mmode\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m'a'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mIOError\u001b[0m: [Errno 2] No such file or directory: '/Users/kanaaax/Google Drive/TS-EUROTRAIN/RESULTS_QSMv3/dataframes/AHBA/ahba_data/donor6.zip'"
     ]
    }
   ],
   "source": [
    "ahba_all = []\n",
    "for donor in donors:\n",
    "    df = return_donor_geneexpression(donor)\n",
    "    ahba_all.append(df)\n",
    "\n",
    "ahba = pd.concat(ahba_all)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#metadata included in the last 13 columns \n",
    "print 'AHBA original dataframe shape', ahba.iloc[:,:-13].shape\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ahba.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Arnatkevic̆iūtė A, Fulcher BD, Fornito A. A practical guide to linking brain-wide gene expression and neuroimaging data. Neuroimage 2019; 189: 353–367.\n",
    "\n",
    "3.1. Step 1. Probe-to-gene re-annotation \n",
    "\n",
    "In microarray experiments, probe sequences correspond to a unique\n",
    "portion of DNA and are assigned to genes based on available genome sequencing databases (O'Leary et al., 2016). While the AHBA (and other platforms) provide annotation tables where probes are mapped to genes, this information gets outdated with each update of the sequencing da- tabases. An accurate probe-to-gene mapping is essential for obtaining biologically meaningful findings. It is therefore necessary to re-assign probes to genes using the most current information available. This re-annotation can be done using several methods and toolboxes, some of which are summarized in Table 1. To the best of our knowledge, only three studies using the AHBA have performed probe-to-gene re-an- notation (Richiardi et al., 2015; Eising et al., 2016; Romero-Garcia et al., 2018b).\n",
    "\n",
    "\n",
    "To investigate how probe-to-gene annotations change over time, we supplied a list of all available 60 bp length AHBA probe sequences (n ¼ 58,692) to the Re-annotator toolkit (Arloth et al., 2015)(Table 1). We found that 45,821 probes (78%) were uniquely annotated to a gene and could be related to an entrez ID - a stable identifier for a gene generated by the Entrez Gene database at the National Center for Biotechnology Information (NCBI). A total of 19% of probes were not mapped to a gene, and just under 3% were mapped to multiple genes and could not be unambiguously annotated. Of the probes that were unambiguously annotated to a gene, 3438 (7:5%) of the annotations differed from those provided by the AHBA: 1287 probes were re-annotated to new genes and 2151 probes that were not previously assigned to any gene in the AHBA could now be annotated. Addi- tionally, 6211 (? 10%) probes in the initial AHBA dataset had an inconsistent gene symbol, ID or gene name information according to the NCBI database (https://www.ncbi.nlm.nih.gov/), as of 5th March 2018. Because of these differences, we recommend obtaining probe-to-gene annotations and retrieving the gene symbol ID and name from the latest version of NCBI (ftp://ftp.ncbi.nlm.nih.gov/ gene/DATA/GENE_INFO/Mammalia/). Hereafter, we present all ana- lyses using this newly re-annotated set of 45,821 probes, correspond- ing to 20,232 unique genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##########################################################################\n",
    "\n",
    "# Function to create 3702x20237 (samples x genes) matrix with reannotation\n",
    "\n",
    "##########################################################################\n",
    "\n",
    "# Reannotation based on re-annotator code as done by:\n",
    "# Arnatkevic̆iūtė A, Fulcher BD, Fornito A. A practical guide to linking brain-wide gene expression and neuroimaging data. \n",
    "# Neuroimage 2019; 189: 353–367. https://figshare.com/articles/AHBAdata/6852911\n",
    "# Arloth J, Bader DM, Röh S, Altmann A. Re-Annotator: Annotation pipeline for microarray probe sequences. PLoS One 2015; 10: 1–13.\n",
    "        \n",
    "reannot = pd.read_csv(os.path.join(datadir,'reannotatedProbes.csv'))#.set_index('ID')\n",
    "reannot['ID'] = reannot['ID'].fillna(\"0\").astype(int)\n",
    "reannot['allenID'] = reannot['allenID'].fillna(\"0\").astype(int)\n",
    "reannot = reannot.set_index('probeNames').sort_index()\n",
    "\n",
    "def return_donor_geneexpression_reannot(donor):\n",
    "    print 'Donor:', donor\n",
    "    #I/O\n",
    "    donor_zip  = zipfile.ZipFile(os.path.join(datadir,'%s.zip'%donor))\n",
    "\n",
    "    #read microarray data. take care indexing.... should have 58692 probes\n",
    "    df_microarray = pd.read_csv(StringIO.StringIO(donor_zip.read('MicroarrayExpression.csv')),\n",
    "                                header= None,index_col =0 )\n",
    "    print '- Microarray dataframe shape: ',df_microarray.shape\n",
    "\n",
    "    # Grab probes df \n",
    "    df_probes = pd.read_csv(StringIO.StringIO(donor_zip.read('Probes.csv')), index_col =0 )\n",
    "    del df_probes.index.name\n",
    "    print '- Probes dataframe shape: ',df_probes.shape\n",
    "\n",
    "    # grab microarray column and add probe name as index... now u have a matrix of probes x wells\n",
    "    df_probes = pd.concat([df_microarray, df_probes['probe_name']],axis=1).set_index('probe_name')\n",
    "    del df_probes.index.name\n",
    "    print '- Probes original-annotation dataframe shape: ',df_probes.shape\n",
    "\n",
    "    # drop unannotated AHBA probes based on re-annotater probes \n",
    "    df_probes = df_probes.drop([i for i in df_probes.index if i not in reannot.index], axis =0).sort_index()\n",
    "    df_probes['GeneReannot'] = reannot['geneNames']\n",
    "    print '- Probes re-annotation dataframe shape: ',df_probes.shape\n",
    "\n",
    "    df_gene_exp = df_probes.groupby('GeneReannot').mean().T.reset_index().drop(['index'],axis=1)\n",
    "    df_gene_exp.columns.name = None\n",
    "    print '- Gene-expression Re-annotation dataframe shape: ',df_gene_exp.shape\n",
    "    print \n",
    "    # Add metadata to dataframe\n",
    "    df_meta = pd.read_csv(StringIO.StringIO(donor_zip.read('SampleAnnot.csv')))\n",
    "    df = pd.concat([df_gene_exp,df_meta],axis=1).set_index('well_id')\n",
    "    del df.index.name\n",
    "\n",
    "    # append donor name\n",
    "    df['donor'] = donor\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "reannot.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ahba_all_reannot = []\n",
    "for donor in donors:\n",
    "    df_reannot = return_donor_geneexpression_reannot(donor)\n",
    "    ahba_all_reannot.append(df_reannot)\n",
    "    \n",
    "ahba_reannot = pd.concat(ahba_all_reannot)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# metadata included in the last 13 columns \n",
    "print 'AHBA re-annotated dataframe shape', ahba_reannot.iloc[:,:-13].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ahba_reannot.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Intensity Based Filtering \n",
    "\n",
    "- The background gene list is provided in Dataset S2 and was calculated by excluding probes which did not exceed the background noise in the AHBA dataset (intensity based filtering), as described by [15]. We used code from [15], with options.probeSelections = ‘maxIntensity’, inline with the maximum intensity approach used to derive our regional gene expression values [16].\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "BACKGROUND_GENES = pd.read_csv(os.path.join(datadir,'background_gene_list.txt'), sep='\\t').GeneSymbol.values\n",
    "ahba_ibf = ahba_reannot.drop([i for i in ahba_reannot.columns if i not in BACKGROUND_GENES], axis =1)\n",
    "\n",
    "print 'AHBA re-annotated and intensity based filtered brain genes dataframe shape', ahba_ibf.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save data \n",
    "#ahba.to_csv(os.path.join(datadir,'AHBA_20737X.csv'))\n",
    "#ahba_reannot.to_csv(os.path.join(datadir,'AHBA_20267.csv'))\n",
    "#ahba_ibf.to_csv(os.path.join(datadir,'AHBA_15745.csv'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Extract gene expression for Striatal Coordinates "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "######################################################\n",
    "# Read Sample data  - QSM stat map t-stat values \n",
    "# T-stat values calculated via FSL-Randomise \n",
    "# Coordinates extracted based on mask \n",
    "# Coords are QCd below by checking AHBA ontology to ensure\n",
    "# that they fall within Striatal subdivisions\n",
    "######################################################\n",
    "\n",
    "ahba_dir     = '/Users/kanaaax/Google Drive/TS-EUROTRAIN/RESULTS_QSMv3/dataframes/AHBA/'\n",
    "MNI = pd.read_csv(os.path.join(ahba_dir,'QSM_TSTATS/MNI_NIFTI_VALUES_permute_10K_OCT2.csv'), index_col = 0 )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/kanaaax/anaconda2/lib/python2.7/site-packages/IPython/core/interactiveshell.py:2714: DtypeWarning: Columns (20763,20764,20765) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    }
   ],
   "source": [
    "######################################################\n",
    "# Read Feature data - QSM stat map t-stat values \n",
    "######################################################\n",
    "\n",
    "ahba         = pd.read_csv(os.path.join(datadir,'AHBA_20737.csv'), index_col=0)#\n",
    "ahba_reannot = pd.read_csv(os.path.join(datadir,'AHBA_20267.csv'), index_col=0)#\n",
    "ahba_ibf     = pd.read_csv(os.path.join(datadir,'AHBA_15745.csv'), index_col=0)#"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "# wells outside the striatal masks \n",
    "drop_wells = [2209, # donor1 claustrum\tclaustrum, right\n",
    "              2233, # donor1 claustrum\tclaustrum, left\n",
    "              2257, # donor1 claustrum\tclaustrum, right\n",
    "              2265, # donor1 claustrum\tclaustrum, left\n",
    "              2249, # donor1 claustrum\tclaustrum, left\n",
    "              #2370, # donor1 claustrum, left\t E\n",
    "              2371, # donor1 claustrum\tclaustrum, right\n",
    "              2379, # donor1 claustrum\tclaustrum, left\n",
    "              2977, # donor1\tclaustrum\tclaustrum, left\n",
    "              5667, # donor2 claustrum\tclaustrum, right\n",
    "              5705, # donor2\tclaustrum\tclaustrum, left\n",
    "              5675, # donor2 claustrum\tclaustrum, left\n",
    "              5769, # donor2\tclaustrum\tclaustrum, left\n",
    "              5898, # donor2 dorsal thalamus\tanterior group of nuclei\n",
    "              11270, # donor3\tcorpus callosum\tNaN \n",
    "              127687121, # donor5\tcorpus callosum\tNaN\n",
    "              147628210, # donor5\tclaustrum\tclaustrum, left\n",
    "              147628277, # donor5 claustrum\tclaustrum, left\n",
    "              156906844, # donor5\tclaustrum\tclaustrum, left\n",
    "              160091500, # donor6 basal forebrain\tseptal nuclei\n",
    "              2370,\n",
    "              126432198, # donor4 parolfactory gyri, left\tfrontal lobe\n",
    "              160091566, #   donor6 perifornical nucleus, left\ttuberal region\n",
    "              160091502, #   donor6 ventromedial hypothalamic nucleus, left\ttuberal region\n",
    "              #159226045, # donor6 striatum//caudate nucleus//body of the caudate nucleus//body of caudate nucleus, left'\n",
    "              #1234 # not in STRM \n",
    "             ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Coords based on masking following non-linear transformation\n",
      "Striatal coords =  103\n",
      "Striatal motor coords =  62\n",
      "Striatal exec coords =  55\n",
      "Striatal limbic coords =  45\n",
      "Striatal motor coords = LL 63\n"
     ]
    }
   ],
   "source": [
    "# check for outlier \n",
    "chi_STR    = MNI['STR_tstat_CP_1mm'].dropna()\n",
    "chi_STR3M  = MNI['STR3_MOTOR_tstat_CP_1mm'].dropna()\n",
    "chi_STR3MLL  = MNI['STR3_MOTOR_tstat_LL_1mm'].dropna()\n",
    "chi_STR3E  = MNI['STR3_EXEC_tstat_CP_1mm'].dropna()\n",
    "chi_STR3L  = MNI['STR3_LIMBIC_tstat_CP_1mm'].dropna()\n",
    "\n",
    "print 'Coords based on masking following non-linear transformation'\n",
    "print 'Striatal coords = ', len(chi_STR)\n",
    "print 'Striatal motor coords = ', len(chi_STR3M)\n",
    "print 'Striatal exec coords = ', len(chi_STR3E)\n",
    "print 'Striatal limbic coords = ', len(chi_STR3L)\n",
    "print 'Striatal motor coords = LL', len(chi_STR3MLL)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1234]\n"
     ]
    }
   ],
   "source": [
    "print [i for i in chi_STR3MLL.index if i not in chi_STR3M]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "88\n",
      "48\n",
      "_____\n",
      "48\n",
      "47\n",
      "40\n",
      "48\n"
     ]
    }
   ],
   "source": [
    "ahba_ontology = ['Ontology','Ontology_4', 'donor']\n",
    "\n",
    "def return_ahba_nucleus_ontology_df(df_ahba, df_chi):\n",
    "    \n",
    "    ahba_nuc = df_ahba.drop([i for i in ahba.index if i not in df_chi.index],axis =0)\n",
    "    ahba_nuc = ahba_nuc.drop([i for i in ahba_nuc.columns if i not in ahba_ontology],axis =1)\n",
    "    ahba_nuc = ahba_nuc.drop([i for i in ahba_nuc.index if i in drop_wells],axis=0)\n",
    "\n",
    "    print len(ahba_nuc)\n",
    "    return ahba_nuc\n",
    "\n",
    "df_ahba_str = return_ahba_nucleus_ontology_df(ahba, chi_STR)\n",
    "df_ahba_str3m = return_ahba_nucleus_ontology_df(ahba, chi_STR3M)\n",
    "print '_____'\n",
    "df_ahba_str3mLL = return_ahba_nucleus_ontology_df(ahba, chi_STR3MLL)\n",
    "df_ahba_str3e = return_ahba_nucleus_ontology_df(ahba, chi_STR3E)\n",
    "df_ahba_str3l = return_ahba_nucleus_ontology_df(ahba, chi_STR3L)\n",
    "df_ahba_str3l = return_ahba_nucleus_ontology_df(ahba, chi_STR3MLL)\n",
    "ind_ontlogy_str3m = df_ahba_str3m[df_ahba_str3m['Ontology'].str.contains('striatum')].index\n",
    "ind_ontlogy_str3m = df_ahba_str3m[df_ahba_str3m['Ontology'].str.contains('striatum')].index\n",
    "ind_ontlogy_str3l = df_ahba_str3l[df_ahba_str3l['Ontology'].str.contains('striatum')].index\n",
    "ind_ontlogy_str3e = df_ahba_str3e[df_ahba_str3e['Ontology'].str.contains('striatum')].index\n",
    "ind_ontology_str  = list(itertools.chain(ind_ontlogy_str3m, ind_ontlogy_str3l, ind_ontlogy_str3e))\n",
    "\n",
    "#drop_wells = list(itertools.chain(drop_wells, [i for i in ahba.index if i not in ind_ontology_str]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>donor</th>\n",
       "      <th>Ontology</th>\n",
       "      <th>Ontology_4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1028</th>\n",
       "      <td>donor1</td>\n",
       "      <td>brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//striatum//putamen//putamen, left</td>\n",
       "      <td>basal ganglia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>977</th>\n",
       "      <td>donor1</td>\n",
       "      <td>brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//striatum//putamen//putamen, left</td>\n",
       "      <td>basal ganglia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>941</th>\n",
       "      <td>donor1</td>\n",
       "      <td>brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//striatum//putamen//putamen, left</td>\n",
       "      <td>basal ganglia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>949</th>\n",
       "      <td>donor1</td>\n",
       "      <td>brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//striatum//caudate nucleus//body of the caudate nucleus//body of caudate nucleus, left</td>\n",
       "      <td>basal ganglia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>933</th>\n",
       "      <td>donor1</td>\n",
       "      <td>brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//striatum//putamen//putamen, left</td>\n",
       "      <td>basal ganglia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>788</th>\n",
       "      <td>donor1</td>\n",
       "      <td>brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//striatum//putamen//putamen, left</td>\n",
       "      <td>basal ganglia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>827</th>\n",
       "      <td>donor1</td>\n",
       "      <td>brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//striatum//putamen//putamen, right</td>\n",
       "      <td>basal ganglia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>780</th>\n",
       "      <td>donor1</td>\n",
       "      <td>brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//striatum//putamen//putamen, left</td>\n",
       "      <td>basal ganglia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1170</th>\n",
       "      <td>donor1</td>\n",
       "      <td>brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//striatum//putamen//putamen, right</td>\n",
       "      <td>basal ganglia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1171</th>\n",
       "      <td>donor1</td>\n",
       "      <td>brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//striatum//putamen//putamen, left</td>\n",
       "      <td>basal ganglia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1242</th>\n",
       "      <td>donor1</td>\n",
       "      <td>brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//striatum//putamen//putamen, right</td>\n",
       "      <td>basal ganglia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4938</th>\n",
       "      <td>donor2</td>\n",
       "      <td>brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//striatum//putamen//putamen, left</td>\n",
       "      <td>basal ganglia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4970</th>\n",
       "      <td>donor2</td>\n",
       "      <td>brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//striatum//caudate nucleus//body of the caudate nucleus//body of caudate nucleus, right</td>\n",
       "      <td>basal ganglia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4961</th>\n",
       "      <td>donor2</td>\n",
       "      <td>brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//striatum//putamen//putamen, right</td>\n",
       "      <td>basal ganglia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4906</th>\n",
       "      <td>donor2</td>\n",
       "      <td>brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//striatum//putamen//putamen, left</td>\n",
       "      <td>basal ganglia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5233</th>\n",
       "      <td>donor2</td>\n",
       "      <td>brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//striatum//putamen//putamen, right</td>\n",
       "      <td>basal ganglia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4914</th>\n",
       "      <td>donor2</td>\n",
       "      <td>brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//striatum//putamen//putamen, left</td>\n",
       "      <td>basal ganglia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5050</th>\n",
       "      <td>donor2</td>\n",
       "      <td>brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//striatum//putamen//putamen, right</td>\n",
       "      <td>basal ganglia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5058</th>\n",
       "      <td>donor2</td>\n",
       "      <td>brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//globus pallidus//globus pallidus, external segment//globus pallidus, external segment, right</td>\n",
       "      <td>basal ganglia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7129</th>\n",
       "      <td>donor2</td>\n",
       "      <td>brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//striatum//caudate nucleus//body of the caudate nucleus//body of caudate nucleus, right</td>\n",
       "      <td>basal ganglia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14169</th>\n",
       "      <td>donor3</td>\n",
       "      <td>brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//globus pallidus//globus pallidus, internal segment//globus pallidus, internal segment, left</td>\n",
       "      <td>basal ganglia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14161</th>\n",
       "      <td>donor3</td>\n",
       "      <td>brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//globus pallidus//globus pallidus, external segment//globus pallidus, external segment, left</td>\n",
       "      <td>basal ganglia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13994</th>\n",
       "      <td>donor3</td>\n",
       "      <td>brain//gray matter//diencephalon//subthalamus//subthalamic nucleus, left</td>\n",
       "      <td>subthalamic nucleus, left</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13594</th>\n",
       "      <td>donor3</td>\n",
       "      <td>brain//gray matter//telencephalon//cerebral nuclei//amygdala//central nucleus//central nucleus, left</td>\n",
       "      <td>amygdala</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13665</th>\n",
       "      <td>donor3</td>\n",
       "      <td>brain//gray matter//telencephalon//cerebral nuclei//amygdala//cortico-medial group//cortico-medial group, left</td>\n",
       "      <td>amygdala</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13905</th>\n",
       "      <td>donor3</td>\n",
       "      <td>brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//striatum//putamen//putamen, left</td>\n",
       "      <td>basal ganglia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126432517</th>\n",
       "      <td>donor4</td>\n",
       "      <td>brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//striatum//putamen//putamen, left</td>\n",
       "      <td>basal ganglia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126432120</th>\n",
       "      <td>donor4</td>\n",
       "      <td>brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//striatum//caudate nucleus//body of the caudate nucleus//body of caudate nucleus, left</td>\n",
       "      <td>basal ganglia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127687979</th>\n",
       "      <td>donor4</td>\n",
       "      <td>brain//gray matter//telencephalon//cerebral nuclei//amygdala//central nucleus//central nucleus, left</td>\n",
       "      <td>amygdala</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127687931</th>\n",
       "      <td>donor4</td>\n",
       "      <td>brain//gray matter//telencephalon//cerebral nuclei//amygdala//cortico-medial group//cortico-medial group, left</td>\n",
       "      <td>amygdala</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127687915</th>\n",
       "      <td>donor4</td>\n",
       "      <td>brain//gray matter//telencephalon//cerebral nuclei//amygdala//basomedial nucleus//basomedial nucleus, left</td>\n",
       "      <td>amygdala</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127687867</th>\n",
       "      <td>donor4</td>\n",
       "      <td>brain//gray matter//telencephalon//cerebral nuclei//basal forebrain//substantia innominata//substantia innominata, left//basal nucleus of meynert, left</td>\n",
       "      <td>basal forebrain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126432136</th>\n",
       "      <td>donor4</td>\n",
       "      <td>brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//striatum//putamen//putamen, left</td>\n",
       "      <td>basal ganglia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126439301</th>\n",
       "      <td>donor4</td>\n",
       "      <td>brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//striatum//putamen//putamen, left</td>\n",
       "      <td>basal ganglia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139379739</th>\n",
       "      <td>donor4</td>\n",
       "      <td>brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//striatum//caudate nucleus//body of the caudate nucleus//body of caudate nucleus, left</td>\n",
       "      <td>basal ganglia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156434555</th>\n",
       "      <td>donor5</td>\n",
       "      <td>brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//striatum//caudate nucleus//body of the caudate nucleus//body of caudate nucleus, left</td>\n",
       "      <td>basal ganglia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149052726</th>\n",
       "      <td>donor5</td>\n",
       "      <td>brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//striatum//putamen//putamen, left</td>\n",
       "      <td>basal ganglia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156434587</th>\n",
       "      <td>donor5</td>\n",
       "      <td>brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//globus pallidus//globus pallidus, external segment//globus pallidus, external segment, left</td>\n",
       "      <td>basal ganglia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156434571</th>\n",
       "      <td>donor5</td>\n",
       "      <td>brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//striatum//putamen//putamen, left</td>\n",
       "      <td>basal ganglia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156906972</th>\n",
       "      <td>donor5</td>\n",
       "      <td>brain//gray matter//telencephalon//cerebral nuclei//basal forebrain//substantia innominata//substantia innominata, left//basal nucleus of meynert, left</td>\n",
       "      <td>basal forebrain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159226077</th>\n",
       "      <td>donor6</td>\n",
       "      <td>brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//globus pallidus//globus pallidus, external segment//globus pallidus, external segment, left</td>\n",
       "      <td>basal ganglia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159226061</th>\n",
       "      <td>donor6</td>\n",
       "      <td>brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//striatum//putamen//putamen, left</td>\n",
       "      <td>basal ganglia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159226107</th>\n",
       "      <td>donor6</td>\n",
       "      <td>brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//striatum//putamen//putamen, left</td>\n",
       "      <td>basal ganglia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159226091</th>\n",
       "      <td>donor6</td>\n",
       "      <td>brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//striatum//caudate nucleus//body of the caudate nucleus//body of caudate nucleus, left</td>\n",
       "      <td>basal ganglia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159439025</th>\n",
       "      <td>donor6</td>\n",
       "      <td>brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//globus pallidus//globus pallidus, external segment//globus pallidus, external segment, left</td>\n",
       "      <td>basal ganglia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160092011</th>\n",
       "      <td>donor6</td>\n",
       "      <td>brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//striatum//putamen//putamen, left</td>\n",
       "      <td>basal ganglia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159438993</th>\n",
       "      <td>donor6</td>\n",
       "      <td>brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//striatum//caudate nucleus//body of the caudate nucleus//body of caudate nucleus, left</td>\n",
       "      <td>basal ganglia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159439119</th>\n",
       "      <td>donor6</td>\n",
       "      <td>brain//gray matter//telencephalon//cerebral nuclei//basal forebrain//substantia innominata//substantia innominata, left//basal nucleus of meynert, left</td>\n",
       "      <td>basal forebrain</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            donor  \\\n",
       "1028       donor1   \n",
       "977        donor1   \n",
       "941        donor1   \n",
       "949        donor1   \n",
       "933        donor1   \n",
       "788        donor1   \n",
       "827        donor1   \n",
       "780        donor1   \n",
       "1170       donor1   \n",
       "1171       donor1   \n",
       "1242       donor1   \n",
       "4938       donor2   \n",
       "4970       donor2   \n",
       "4961       donor2   \n",
       "4906       donor2   \n",
       "5233       donor2   \n",
       "4914       donor2   \n",
       "5050       donor2   \n",
       "5058       donor2   \n",
       "7129       donor2   \n",
       "14169      donor3   \n",
       "14161      donor3   \n",
       "13994      donor3   \n",
       "13594      donor3   \n",
       "13665      donor3   \n",
       "13905      donor3   \n",
       "126432517  donor4   \n",
       "126432120  donor4   \n",
       "127687979  donor4   \n",
       "127687931  donor4   \n",
       "127687915  donor4   \n",
       "127687867  donor4   \n",
       "126432136  donor4   \n",
       "126439301  donor4   \n",
       "139379739  donor4   \n",
       "156434555  donor5   \n",
       "149052726  donor5   \n",
       "156434587  donor5   \n",
       "156434571  donor5   \n",
       "156906972  donor5   \n",
       "159226077  donor6   \n",
       "159226061  donor6   \n",
       "159226107  donor6   \n",
       "159226091  donor6   \n",
       "159439025  donor6   \n",
       "160092011  donor6   \n",
       "159438993  donor6   \n",
       "159439119  donor6   \n",
       "\n",
       "                                                                                                                                                                  Ontology  \\\n",
       "1028                                                                   brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//striatum//putamen//putamen, left   \n",
       "977                                                                    brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//striatum//putamen//putamen, left   \n",
       "941                                                                    brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//striatum//putamen//putamen, left   \n",
       "949               brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//striatum//caudate nucleus//body of the caudate nucleus//body of caudate nucleus, left   \n",
       "933                                                                    brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//striatum//putamen//putamen, left   \n",
       "788                                                                    brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//striatum//putamen//putamen, left   \n",
       "827                                                                   brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//striatum//putamen//putamen, right   \n",
       "780                                                                    brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//striatum//putamen//putamen, left   \n",
       "1170                                                                  brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//striatum//putamen//putamen, right   \n",
       "1171                                                                   brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//striatum//putamen//putamen, left   \n",
       "1242                                                                  brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//striatum//putamen//putamen, right   \n",
       "4938                                                                   brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//striatum//putamen//putamen, left   \n",
       "4970             brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//striatum//caudate nucleus//body of the caudate nucleus//body of caudate nucleus, right   \n",
       "4961                                                                  brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//striatum//putamen//putamen, right   \n",
       "4906                                                                   brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//striatum//putamen//putamen, left   \n",
       "5233                                                                  brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//striatum//putamen//putamen, right   \n",
       "4914                                                                   brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//striatum//putamen//putamen, left   \n",
       "5050                                                                  brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//striatum//putamen//putamen, right   \n",
       "5058       brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//globus pallidus//globus pallidus, external segment//globus pallidus, external segment, right   \n",
       "7129             brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//striatum//caudate nucleus//body of the caudate nucleus//body of caudate nucleus, right   \n",
       "14169       brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//globus pallidus//globus pallidus, internal segment//globus pallidus, internal segment, left   \n",
       "14161       brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//globus pallidus//globus pallidus, external segment//globus pallidus, external segment, left   \n",
       "13994                                                                                             brain//gray matter//diencephalon//subthalamus//subthalamic nucleus, left   \n",
       "13594                                                                 brain//gray matter//telencephalon//cerebral nuclei//amygdala//central nucleus//central nucleus, left   \n",
       "13665                                                       brain//gray matter//telencephalon//cerebral nuclei//amygdala//cortico-medial group//cortico-medial group, left   \n",
       "13905                                                                  brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//striatum//putamen//putamen, left   \n",
       "126432517                                                              brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//striatum//putamen//putamen, left   \n",
       "126432120         brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//striatum//caudate nucleus//body of the caudate nucleus//body of caudate nucleus, left   \n",
       "127687979                                                             brain//gray matter//telencephalon//cerebral nuclei//amygdala//central nucleus//central nucleus, left   \n",
       "127687931                                                   brain//gray matter//telencephalon//cerebral nuclei//amygdala//cortico-medial group//cortico-medial group, left   \n",
       "127687915                                                       brain//gray matter//telencephalon//cerebral nuclei//amygdala//basomedial nucleus//basomedial nucleus, left   \n",
       "127687867          brain//gray matter//telencephalon//cerebral nuclei//basal forebrain//substantia innominata//substantia innominata, left//basal nucleus of meynert, left   \n",
       "126432136                                                              brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//striatum//putamen//putamen, left   \n",
       "126439301                                                              brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//striatum//putamen//putamen, left   \n",
       "139379739         brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//striatum//caudate nucleus//body of the caudate nucleus//body of caudate nucleus, left   \n",
       "156434555         brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//striatum//caudate nucleus//body of the caudate nucleus//body of caudate nucleus, left   \n",
       "149052726                                                              brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//striatum//putamen//putamen, left   \n",
       "156434587   brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//globus pallidus//globus pallidus, external segment//globus pallidus, external segment, left   \n",
       "156434571                                                              brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//striatum//putamen//putamen, left   \n",
       "156906972          brain//gray matter//telencephalon//cerebral nuclei//basal forebrain//substantia innominata//substantia innominata, left//basal nucleus of meynert, left   \n",
       "159226077   brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//globus pallidus//globus pallidus, external segment//globus pallidus, external segment, left   \n",
       "159226061                                                              brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//striatum//putamen//putamen, left   \n",
       "159226107                                                              brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//striatum//putamen//putamen, left   \n",
       "159226091         brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//striatum//caudate nucleus//body of the caudate nucleus//body of caudate nucleus, left   \n",
       "159439025   brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//globus pallidus//globus pallidus, external segment//globus pallidus, external segment, left   \n",
       "160092011                                                              brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//striatum//putamen//putamen, left   \n",
       "159438993         brain//gray matter//telencephalon//cerebral nuclei//basal ganglia//striatum//caudate nucleus//body of the caudate nucleus//body of caudate nucleus, left   \n",
       "159439119          brain//gray matter//telencephalon//cerebral nuclei//basal forebrain//substantia innominata//substantia innominata, left//basal nucleus of meynert, left   \n",
       "\n",
       "                          Ontology_4  \n",
       "1028                   basal ganglia  \n",
       "977                    basal ganglia  \n",
       "941                    basal ganglia  \n",
       "949                    basal ganglia  \n",
       "933                    basal ganglia  \n",
       "788                    basal ganglia  \n",
       "827                    basal ganglia  \n",
       "780                    basal ganglia  \n",
       "1170                   basal ganglia  \n",
       "1171                   basal ganglia  \n",
       "1242                   basal ganglia  \n",
       "4938                   basal ganglia  \n",
       "4970                   basal ganglia  \n",
       "4961                   basal ganglia  \n",
       "4906                   basal ganglia  \n",
       "5233                   basal ganglia  \n",
       "4914                   basal ganglia  \n",
       "5050                   basal ganglia  \n",
       "5058                   basal ganglia  \n",
       "7129                   basal ganglia  \n",
       "14169                  basal ganglia  \n",
       "14161                  basal ganglia  \n",
       "13994      subthalamic nucleus, left  \n",
       "13594                       amygdala  \n",
       "13665                       amygdala  \n",
       "13905                  basal ganglia  \n",
       "126432517              basal ganglia  \n",
       "126432120              basal ganglia  \n",
       "127687979                   amygdala  \n",
       "127687931                   amygdala  \n",
       "127687915                   amygdala  \n",
       "127687867            basal forebrain  \n",
       "126432136              basal ganglia  \n",
       "126439301              basal ganglia  \n",
       "139379739              basal ganglia  \n",
       "156434555              basal ganglia  \n",
       "149052726              basal ganglia  \n",
       "156434587              basal ganglia  \n",
       "156434571              basal ganglia  \n",
       "156906972            basal forebrain  \n",
       "159226077              basal ganglia  \n",
       "159226061              basal ganglia  \n",
       "159226107              basal ganglia  \n",
       "159226091              basal ganglia  \n",
       "159439025              basal ganglia  \n",
       "160092011              basal ganglia  \n",
       "159438993              basal ganglia  \n",
       "159439119            basal forebrain  "
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_ahba_str3m"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Creating n_samples x n_features dataframes for: STR3_MOTOR_tstat_CP_1mm\n",
      " - Nucleus contains this many samples: 48\n",
      "\n",
      "Creating n_samples x n_features dataframes for: STR3_EXEC_tstat_CP_1mm\n",
      " - Nucleus contains this many samples: 47\n",
      "\n",
      "Creating n_samples x n_features dataframes for: STR3_LIMBIC_tstat_CP_1mm\n",
      " - Nucleus contains this many samples: 40\n",
      "\n",
      "Creating n_samples x n_features dataframes for: STR3_MOTOR_tstat_LL_1mm\n",
      " - Nucleus contains this many samples: 48\n",
      "\n"
     ]
    }
   ],
   "source": [
    "def make_sample_feature_dataframe(nucleus): \n",
    "    \n",
    "    print 'Creating n_samples x n_features dataframes for:', nucleus\n",
    "    # Grab t-stat dataframe.... contains t-stat values for wells within nucleus \n",
    "    \n",
    "    chi = pd.DataFrame(MNI[nucleus]).dropna().sort_index()\n",
    "    chi = chi.drop([i for i in chi.index if i in drop_wells],axis=0)\n",
    "    print ' - Nucleus contains this many samples:', len(chi.index)    \n",
    "    print ''\n",
    "    \n",
    "    ahbax = ahba.iloc[:,:-28].sort_index().drop([i for i in ahba.index if i not in chi.index ], axis =0).sort_index()\n",
    "    ahba_reannotx = ahba_reannot.iloc[:,:-13].sort_index().drop([i for i in ahba.index if i not in chi.index ], axis =0).sort_index()\n",
    "    ahba_ibfx = ahba_ibf.sort_index().drop([i for i in ahba.index if i not in chi.index ], axis =0).sort_index()\n",
    "    \n",
    "    ahbax.to_csv(os.path.join(datadir,'index/AHBA_20737_index_%s.csv'%nucleus))\n",
    "    ahba_reannotx.to_csv(os.path.join(datadir,'index/AHBA_20267_index_%s.csv'%nucleus))\n",
    "    ahba_ibfx.to_csv(os.path.join(datadir,'index/AHBA_15745_index_%s.csv'%nucleus))\n",
    "    \n",
    "    chi.to_csv(os.path.join(datadir,'index/chi_index_%s.csv'%nucleus))\n",
    "    \n",
    "make_sample_feature_dataframe('STR3_MOTOR_tstat_CP_1mm')\n",
    "make_sample_feature_dataframe('STR3_EXEC_tstat_CP_1mm')\n",
    "make_sample_feature_dataframe('STR3_LIMBIC_tstat_CP_1mm')\n",
    "#make_sample_feature_dataframe('STR_tstat_CP_1mm')\n",
    "make_sample_feature_dataframe('STR3_MOTOR_tstat_LL_1mm')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "#chi.to_csv(os.path.join(datadir,'index/chi_index_%s.csv'%nucleus))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Creating n_samples x n_features dataframes for full striatum\n",
      "(48, 1)\n",
      "(38, 1)\n",
      "(36, 1)\n",
      " - Nucleus contains this many samples: 122\n",
      "\n",
      "(122, 20737)\n",
      "(122, 20267)\n",
      "(122, 15745)\n"
     ]
    }
   ],
   "source": [
    "def make_str_combined_sample_feature_dataframe():\n",
    "    \n",
    "    print 'Creating n_samples x n_features dataframes for full striatum'\n",
    "    \n",
    "    # Grab t-stat dataframe.... contains t-stat values for wells within nucleus \n",
    "    chiM = pd.DataFrame(MNI['STR3_MOTOR_tstat_CP_1mm']).dropna()\n",
    "    chiM = chiM.drop([i for i in chiM.index if i in drop_wells],axis=0)\n",
    "    print chiM.shape\n",
    "\n",
    "    chiL = pd.DataFrame(MNI['STR3_LIMBIC_tstat_CP_1mm']).dropna()\n",
    "    chiL = chiL.drop([i for i in chiL.index if i in drop_wells],axis=0)\n",
    "    chiL = chiL.drop([i for i in chiL.index if i in chiM.index],axis=0)\n",
    "    print chiL.shape\n",
    "    \n",
    "    chiE = pd.DataFrame(MNI['STR3_EXEC_tstat_CP_1mm']).dropna()\n",
    "    chiE = chiE.drop([i for i in chiE.index if i in drop_wells],axis=0)\n",
    "    chiE = chiE.drop([i for i in chiE.index if i in chiM.index],axis=0)\n",
    "    chiE = chiE.drop([i for i in chiE.index if i in chiL.index],axis=0)\n",
    "    print chiE.shape\n",
    "\n",
    "    chiM.rename(columns={'STR3_MOTOR_tstat_CP_1mm':'STR3'}, inplace=True)\n",
    "    chiE.rename(columns={'STR3_EXEC_tstat_CP_1mm':'STR3'}, inplace=True)\n",
    "    chiL.rename(columns={'STR3_LIMBIC_tstat_CP_1mm':'STR3'}, inplace=True)\n",
    "    \n",
    "    chi = pd.concat([chiM,chiE, chiL],axis = 0)\n",
    "    print ' - Nucleus contains this many samples:', len(chi.index)    \n",
    "    print ''\n",
    "    \n",
    "    ahbax = ahba.iloc[:,:-28].sort_index().drop([i for i in ahba.index if i not in chi.index ], axis =0).sort_index()\n",
    "    ahba_reannotx = ahba_reannot.iloc[:,:-13].sort_index().drop([i for i in ahba.index if i not in chi.index ], axis =0).sort_index()\n",
    "    ahba_ibfx = ahba_ibf.sort_index().drop([i for i in ahba.index if i not in chi.index ], axis =0).sort_index()\n",
    "    \n",
    "    nucleus = 'STR3_tstat_CP_1mm'\n",
    "    \n",
    "    ahbax.to_csv(os.path.join(datadir,'index/AHBA_20737_index_%s.csv'%nucleus))\n",
    "    ahba_reannotx.to_csv(os.path.join(datadir,'index/AHBA_20267_index_%s.csv'%nucleus))\n",
    "    ahba_ibfx.to_csv(os.path.join(datadir,'index/AHBA_15745_index_%s.csv'%nucleus))\n",
    "    \n",
    "    chi.to_csv(os.path.join(datadir,'index/chi_index_%s.csv'%nucleus))\n",
    "\n",
    "    print ahbax.shape\n",
    "    print ahba_reannotx.shape\n",
    "    print ahba_ibfx.shape\n",
    "    \n",
    "    return chi\n",
    "x = make_str_combined_sample_feature_dataframe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "122"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "48+36+38"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(48, 20737)\n",
      "(48, 20267)\n",
      "(48, 15745)\n",
      "\n",
      "(47, 20737)\n",
      "(47, 20267)\n",
      "(47, 15745)\n",
      "\n",
      "(40, 20737)\n",
      "(40, 20267)\n",
      "(40, 15745)\n"
     ]
    }
   ],
   "source": [
    "print pd.read_csv(os.path.join(datadir,'index/AHBA_20737_index_STR3_MOTOR_tstat_CP_1mm.csv'), index_col=0).shape \n",
    "print pd.read_csv(os.path.join(datadir,'index/AHBA_20267_index_STR3_MOTOR_tstat_CP_1mm.csv'), index_col=0).shape \n",
    "print pd.read_csv(os.path.join(datadir,'index/AHBA_15745_index_STR3_MOTOR_tstat_CP_1mm.csv'), index_col=0).shape \n",
    "print ''\n",
    "print pd.read_csv(os.path.join(datadir,'index/AHBA_20737_index_STR3_EXEC_tstat_CP_1mm.csv'), index_col=0).shape \n",
    "print pd.read_csv(os.path.join(datadir,'index/AHBA_20267_index_STR3_EXEC_tstat_CP_1mm.csv'), index_col=0).shape \n",
    "print pd.read_csv(os.path.join(datadir,'index/AHBA_15745_index_STR3_EXEC_tstat_CP_1mm.csv'), index_col=0).shape \n",
    "print ''\n",
    "print pd.read_csv(os.path.join(datadir,'index/AHBA_20737_index_STR3_LIMBIC_tstat_CP_1mm.csv'), index_col=0).shape \n",
    "print pd.read_csv(os.path.join(datadir,'index/AHBA_20267_index_STR3_LIMBIC_tstat_CP_1mm.csv'), index_col=0).shape \n",
    "print pd.read_csv(os.path.join(datadir,'index/AHBA_15745_index_STR3_LIMBIC_tstat_CP_1mm.csv'), index_col=0).shape \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(122, 20737)\n",
      "(122, 20267)\n",
      "(122, 15745)\n"
     ]
    }
   ],
   "source": [
    "print pd.read_csv(os.path.join(datadir,'index/AHBA_20737_index_STR3_tstat_CP_1mm.csv'), index_col=0).shape \n",
    "print pd.read_csv(os.path.join(datadir,'index/AHBA_20267_index_STR3_tstat_CP_1mm.csv'), index_col=0).shape \n",
    "print pd.read_csv(os.path.join(datadir,'index/AHBA_15745_index_STR3_tstat_CP_1mm.csv'), index_col=0).shape \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
